livestream/event_types.go

```diff
@@ -0,0 +1,102 @@
+package livestream
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+type (
+	// EventType is the type of a livestream_events record.
+	// This type must stay in sync with the livestream_event_type enum in the database.
+	EventType        string
+	EventWithoutData struct {
+		ID   int       `db:"livestream_event_id" json:"livestreamEventID"`
+		Type EventType `db:"event_type" json:"type"`
+		Time time.Time `db:"event_time" json:"time"`
+	}
+	// Event is a livestream_events record.
+	Event struct {
+		EventWithoutData
+		Data EventPayload `db:"event_data" json:"data"`
+	}
+)
+
+const (
+	EventStarted        EventType = "started"
+	EventEnded          EventType = "ended"
+	EventLinked         EventType = "linked"
+	EventUnlinked       EventType = "unlinked"
+	EventStreamReceived EventType = "stream_received"
+	EventStreamLost     EventType = "stream_lost"
+	EventError          EventType = "error"
+)
+
+// EventPayload is the type of all livestream event payloads, used only for type checking.
+type EventPayload interface {
+	isEventPayload()
+}
+
+// UnmarshalEventPayload unmarshals a JSON event payload into the appropriate type.
+func UnmarshalEventPayload(typ EventType, raw json.RawMessage) (EventPayload, error) {
+	var data EventPayload
+	switch typ {
+	case EventStarted:
+		data = &EventStartedPayload{}
+	case EventEnded:
+		data = &EventEndedPayload{}
+	case EventLinked:
+		data = &EventLinkedPayload{}
+	case EventUnlinked:
+		data = &EventUnlinkedPayload{}
+	case EventStreamReceived:
+		data = &EventStreamReceivedPayload{}
+	case EventStreamLost:
+		data = &EventStreamLostPayload{}
+	case EventError:
+		data = &EventErrorPayload{}
```

Comment:
What is the benefit of an `EventType` and being split this way? Although I might be able to understand it a bit more with some time

Reply:
A `livestream_event event` record has two fields, `event_type`, a string, and `event_data`, which is raw JSON. The benefit of separating the two is that unmarshaling the `event_data` can be type-safe - we can inspect the `event_type` to determine what type the payload is, instead of unmarshaling to a `map[string]any` or similar.

Reply:
Yeah that's understandable